core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala

```diff
@@ -144,31 +259,55 @@ private[zio] final class FiberContext[E, A](platform: Platform, startEnv: AnyRef
                     case ZIO.Tags.Succeed =>
                       val io2 = nested.asInstanceOf[ZIO.Succeed[Any]]
 
-                      curIo = io.k(io2.value)
+                      if (traceExec && inTracingRegion) addTrace(k)
+
+                      curIo = k(io2.value)
 
                     case ZIO.Tags.EffectTotal =>
-                      val io2 = nested.asInstanceOf[ZIO.EffectTotal[Any]]
+                      val io2    = nested.asInstanceOf[ZIO.EffectTotal[Any]]
+                      val effect = io2.effect
+
+                      val kTrace = fastPathTrace(k, effect)
 
-                      curIo = io.k(io2.effect())
+                      val value = effect()
+
+                      // delete continuation as it was "popped" after success
+                      if (traceStack && (kTrace ne null)) fastPathFlatMapContinuationTrace = null
+                      // record continuation in exec as we're just "passing" it
+                      if (traceExec && (kTrace ne null)) execTrace.put(kTrace)
+
+                      curIo = k(value)
 
                     case ZIO.Tags.EffectPartial =>
-                      val io2 = nested.asInstanceOf[ZIO.EffectPartial[Any]]
+                      val io2    = nested.asInstanceOf[ZIO.EffectPartial[Any]]
+                      val effect = io2.effect
 
-                      var nextIo = null.asInstanceOf[IO[E, Any]]
-                      val value = try io2.effect()
+                      val kTrace = fastPathTrace(k, effect)
+
+                      var failIO = null.asInstanceOf[IO[E, Any]]
+                      val value = try effect()
                       catch {
                         case t: Throwable if !platform.fatal(t) =>
-                          nextIo = ZIO.fail(t.asInstanceOf[E])
+                          failIO = ZIO.fail(t.asInstanceOf[E])
+                      }
+
+                      if (failIO eq null) {
+                        // delete continuation as it was "popped" after success
+                        if (traceStack && (kTrace ne null)) fastPathFlatMapContinuationTrace = null
+                        // record continuation in exec as we're just "passing" it
+                        if (traceExec && (kTrace ne null)) execTrace.put(kTrace)
+
+                        curIo = k(value)
+                      } else {
+                        curIo = failIO
                       }
-                      if (nextIo eq null) curIo = io.k(value)
-                      else curIo = nextIo
 
                     case _ =>
                       // Fallback case. We couldn't evaluate the LHS so we have to
                       // use the stack:
                       curIo = nested
-
-                      stack.push(io.k)
+                      if (traceStack && inTracingRegion) addStackTrace(k)
+                      stack.push(k)
```

Comment:
After seeing this pattern repeated multiple times, I'm more convinced we need a utility method that can call `addStackTrace(k)` and `stack.push(k)` at the same time.

Reply:
:+1: